core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala

```diff
@@ -106,6 +108,28 @@ private[spark] class HashShuffleWriter[K, V](
       writer.commitAndClose()
       writer.fileSegment().length
     }
+    // rename all shuffle files to final paths
+    // Note: there is only one ShuffleBlockResolver in executor
+    shuffleBlockResolver.synchronized {
+      shuffle.writers.zipWithIndex.foreach { case (writer, i) =>
+        val output = blockManager.diskBlockManager.getFile(writer.blockId)
+        if (sizes(i) > 0) {
+          if (output.exists()) {
+            // update the size of output for MapStatus
+            sizes(i) = output.length()
+            writer.file.delete()
+          } else {
+            if (!writer.file.renameTo(output)) {
```

Comment:
`// Commit by renaming our temporary file to something the shuffle reader expects`